browser/brave_ads/creatives/search_result_ad/creative_search_result_ad_tab_helper.cc

```diff
@@ -0,0 +1,263 @@
+  return profile->GetPrefs()->GetBoolean(brave_rewards::prefs::kEnabled);
+}
+
+void CreativeSearchResultAdTabHelper::MaybeTriggerCreativeAdClickedEvent(
+    const GURL& url) {
+  if (creative_search_result_ad_handler_) {
+    creative_search_result_ad_handler_->MaybeTriggerCreativeAdClickedEvent(url);
+  }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+AdsService* CreativeSearchResultAdTabHelper::GetAdsService() const {
+  if (g_ads_service_for_testing) {
+    CHECK_IS_TEST();
+
+    return g_ads_service_for_testing;
+  }
+
+  Profile* const profile =
+      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
+  return AdsServiceFactory::GetForProfile(profile);
+}
+
+void CreativeSearchResultAdTabHelper::MaybeCreateCreativeSearchResultAdHandler(
+    content::NavigationHandle* navigation_handle) {
+  CHECK(navigation_handle);
+
+  if (!ShouldHandleCreativeAdEvents()) {
+    return;
+  }
+
+  // Do not trigger ad viewed events if the user navigates back or forward.
+  const bool should_trigger_creative_ad_viewed_events =
+      (navigation_handle->GetPageTransition() &
+       ui::PAGE_TRANSITION_FORWARD_BACK) == 0;
+
+  creative_search_result_ad_handler_ =
+      CreativeSearchResultAdHandler::MaybeCreate(
+          GetAdsService(), navigation_handle->GetURL(),
+          should_trigger_creative_ad_viewed_events);
+}
+
+void CreativeSearchResultAdTabHelper::
+    MaybeExtractCreativeAdPlacementIdsFromWebPageAndHandleViewedEvents() {
+  if (!ShouldHandleCreativeAdEvents()) {
+    return;
+  }
+
+  if (!creative_search_result_ad_handler_) {
+    return;
+  }
+
+  creative_search_result_ad_handler_
+      ->MaybeExtractCreativeAdPlacementIdsFromWebPage(
+          web_contents()->GetPrimaryMainFrame(),
+          base::BindOnce(&CreativeSearchResultAdTabHelper::
+                             MaybeHandleCreativeAdViewedEvents,
+                         weak_factory_.GetWeakPtr()));
+}
+
+void CreativeSearchResultAdTabHelper::MaybeHandleCreativeAdViewedEvents(
+    std::vector<std::string> placement_ids) {
+  for (const auto& placement_id : placement_ids) {
+    MaybeHandleCreativeAdViewedEvent(placement_id);
+  }
+}
+
+void CreativeSearchResultAdTabHelper::MaybeHandleCreativeAdViewedEvent(
+    const std::string& placement_id) {
+  CHECK(!placement_id.empty());
+
+  const std::string javascript = base::ReplaceStringPlaceholders(
+      kDataPlacementIdVisibilityCheckJavaScript, {placement_id}, nullptr);
+
+  web_contents()->GetPrimaryMainFrame()->ExecuteJavaScriptInIsolatedWorld(
+      base::UTF8ToUTF16(javascript),
+      base::BindOnce(&CreativeSearchResultAdTabHelper::
+                         MaybeHandleCreativeAdViewedEventCallback,
+                     weak_factory_.GetWeakPtr(), placement_id),
+      ISOLATED_WORLD_ID_BRAVE_INTERNAL);
+}
+
+void CreativeSearchResultAdTabHelper::MaybeHandleCreativeAdViewedEventCallback(
+    const std::string& placement_id,
+    base::Value value) {
```

Comment:
can we add const here: `const base::Value value`

Reply:
Fixed